daemon/image_delete.go

```diff
@@ -232,7 +232,9 @@ func (daemon *Daemon) getContainerUsingImage(imageID image.ID) *container.Contai
 // optional tag or digest reference. If tag or digest is omitted, the default
 // tag is used. Returns the resolved image reference and an error.
 func (daemon *Daemon) removeImageRef(ref reference.Named) (reference.Named, error) {
-	ref = reference.WithDefaultTag(ref)
+	if reference.IsNameOnly(ref) {
+		ref = reference.EnsureTagged(ref)
+	}
```

Comment:
This is kind of an annoying one, don't really want to make helper functions for the different criteria. Even though at no point do we use ensuretagged when it could be adding to a reference which already has a digest. This is more lines of code, but I think it reads much nicer since it is more obvious how and when the reference is being transformed, rather than having a helper function which reads well enough to understand what it is doing and for the appropriate combinations.